File: packages/vm/wasmlib/ts/wasmlib/events.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {ScAddress, ScAgentID, ScChainID, ScColor, ScHash, ScHname, ScRequestID} from "./hashtypes";
import * as keys from "./keys";
import {base58Encode, ROOT} from "./context";

// encodes separate entities into a byte buffer
export class EventEncoder {
    event: string;

    // constructs an encoder
    constructor(eventName: string) {
        this.event = eventName;
        let timestamp = ROOT.getInt64(keys.KEY_TIMESTAMP).value();
        this.int64(timestamp / 1_000_000_000);
    }

    // encodes an ScAddress into the byte buffer
    address(value: ScAddress): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an ScAgentID into the byte buffer
    agentID(value: ScAgentID): EventEncoder {
        return this.string(value.toString());
    }

    // encodes a Bool into the byte buffer
    bool(value: bool): EventEncoder {
        return this.uint8(value ? 1 : 0);
    }

    // encodes a substring of bytes into the byte buffer
    bytes(value: u8[]): EventEncoder {
        return this.string(base58Encode(value));
    }

    // encodes an ScChainID into the byte buffer
    chainID(value: ScChainID): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an ScColor into the byte buffer
    color(value: ScColor): EventEncoder {
        return this.string(value.toString());
    }

    // retrieve the encoded byte buffer
    emit(): void {
        ROOT.getString(keys.KEY_EVENT).setValue(this.event);
    }

    // encodes an ScHash into the byte buffer
    hash(value: ScHash): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an ScHname into the byte buffer
    hname(value: ScHname): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Int8 into the byte buffer
    int8(value: i8): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Int16 into the byte buffer
    int16(value: i16): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Int32 into the byte buffer
    int32(value: i32): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Int64 into the byte buffer
    int64(value: i64): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an ScRequestID into the byte buffer
    requestID(value: ScRequestID): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an UTF-8 text string into the byte buffer
    string(value: string): EventEncoder {
        //TODO encode potential vertical bars that are present in the value string
        this.event += "|" + value;
        return this;
    }

    // encodes an Uint8 into the byte buffer
    uint8(value: u8): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Uint16 into the byte buffer
    uint16(value: u16): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Uint32 into the byte buffer
    uint32(value: u32): EventEncoder {
        return this.string(value.toString());
    }

    // encodes an Uint64 into the byte buffer
    uint64(value: u64): EventEncoder {
        return this.string(value.toString());
    }
}

File: packages/vm/wasmlib/go/wasmlib/events.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"strconv"

	"github.com/iotaledger/wasp/packages/vm/wasmlib/go/wasmlib/wasmtypes"
)

// encodes separate entities into a byte buffer
type EventEncoder struct {
	event string
}

func NewEventEncoder(eventName string) *EventEncoder {
	e := &EventEncoder{event: eventName}
	timestamp := ScSandbox{}.Timestamp()
	// convert nanoseconds to seconds
	e.Encode(wasmtypes.Uint64ToString(timestamp / 1_000_000_000))
	return e
}

func (e *EventEncoder) Address(value wasmtypes.ScAddress) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) AgentID(value wasmtypes.ScAgentID) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) Bool(value bool) *EventEncoder {
	if value {
		return e.Uint8(1)
	}
	return e.Uint8(0)
}

func (e *EventEncoder) Bytes(value []byte) *EventEncoder {
	return e.String(base58Encode(value))
}

func (e *EventEncoder) ChainID(value wasmtypes.ScChainID) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) Color(value wasmtypes.ScColor) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) Emit() {
	ScSandboxFunc{}.Event(e.event)
}

func (e *EventEncoder) Encode(value string) {
	// TODO encode potential vertical bars that are present in the value string
	e.event += "|" + value
}

func (e *EventEncoder) Hash(value wasmtypes.ScHash) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) Hname(value wasmtypes.ScHname) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) Int8(value int8) *EventEncoder {
	return e.Int64(int64(value))
}

func (e *EventEncoder) Int16(value int16) *EventEncoder {
	return e.Int64(int64(value))
}

func (e *EventEncoder) Int32(value int32) *EventEncoder {
	return e.Int64(int64(value))
}

func (e *EventEncoder) Int64(value int64) *EventEncoder {
	return e.String(strconv.FormatInt(value, 10))
}

func (e *EventEncoder) RequestID(value wasmtypes.ScRequestID) *EventEncoder {
	return e.String(value.String())
}

func (e *EventEncoder) String(value string) *EventEncoder {
	// TODO encode potential vertical bars that are present in the value string
	e.event += "|" + value
	return e
}

func (e *EventEncoder) Uint8(value uint8) *EventEncoder {
	return e.Uint64(uint64(value))
}

func (e *EventEncoder) Uint16(value uint16) *EventEncoder {
	return e.Uint64(uint64(value))
}

func (e *EventEncoder) Uint32(value uint32) *EventEncoder {
	return e.Uint64(uint64(value))
}

func (e *EventEncoder) Uint64(value uint64) *EventEncoder {
	return e.String(strconv.FormatUint(value, 10))
}
